client/client.go

```diff
@@ -267,11 +274,35 @@ func (cli *Client) Close() error {
 // we cannot do the negotiation lazily.
 func (cli *Client) checkVersion(ctx context.Context) error {
 	if !cli.manualOverride && cli.negotiateVersion && !cli.negotiated {
-		ping, err := cli.Ping(ctx)
+		// Acquire the negotiateLock to ensure exclusive access to versionNegotiator.
+		// This prevents concurrent negotiations.
+		cli.negotiateLock.Lock()
+		defer cli.negotiateLock.Unlock()
+
+		if cli.versionNegotiator == nil {
+			// The current context is passed to the Ping method to respect any deadlines or cancellations.
+			cli.versionNegotiator = sync.OnceValue(func() error {
+				ping, err := cli.Ping(ctx)
+				if err != nil {
+					return err
+				}
+
+				cli.negotiateAPIVersionPing(ping)
+
+				return nil
+			})
+		}
+
+		err := cli.versionNegotiator()
```

Comment:
`cli.versionNegotiator()` is only ever called inside a `cli.negotiateLock` critical section, meaning at most one goroutine can call it at a time. So there is no need for the synchronization behavior of `sync.OnceValue`. And there is no need to remember the error returned by `cli.Ping()` because, if the ping is successful, `cli.negotiated` gets set to `true` by `cli.negotiateAPIVersionPing()`. It looks to me like `sync.OnceValue` is not useful in this situation; adding complexity without providing any benefit in return.

Reply:
I got to this approach due to #42379; if you allow two functions to enter this part, they may "queue up" to perform the same ping twice. However, I see this can be worked around by extending the critical zone to this function in its entirety (this way, cli.negotiated will have already been set by the next time this is reached).

I was trying to avoid impacting the performance characteristics of this function after the first call since it's called by nearly every *Client method, and I do think this is the only way to get around a mutex for the entire function, but I see that wrapping it in a mutex is clearly the cleaner / more maintainable solution

Reply:
Alternatively, you could check for cli.negotiated again and early return right after entering the critical zone, but this is very confusing

Reply:
Upon further inspection, `checkVersion` is called in `NewVersionError`, which means it's called 5 times in `ContainerCreate` alone , twice in `ImageBuild`, and some more times in many other places. Entering a critical zone 5 times every time a container is created doesn't seem optimal either